src/uuid/string.cr

```diff
@@ -0,0 +1,89 @@
+struct UUID
+  # Raises `ArgumentError` if string `value` at index `i` doesn't contain hex digit followed by another hex digit.
+  # TODO: Move to String#digits?(base, offset, size) or introduce strict String#[index, size].to_u8(base)! which doesn't
+  #       allow non-digits. The problem it solves is that " 1".to_u8(16) is fine but if it appears inside hexstring
+  #       it's not correct and there should be stdlib function to support it, without a need to build this kind of
+  #       helpers.
+  def self.string_has_hex_pair_at!(value : String, i)
+    unless value[i].hex? && value[i + 1].hex?
+      raise ArgumentError.new [
+        "Invalid hex character at position #{i * 2} or #{i * 2 + 1}",
+        "expected '0' to '9', 'a' to 'f' or 'A' to 'F'.",
+      ].join(", ")
+    end
+  end
+
+  # Creates new UUID by decoding `value` string from hyphenated (ie. `ba714f86-cac6-42c7-8956-bcf5105e1b81`),
+  # hexstring (ie. `89370a4ab66440c8add39e06f2bb6af6`) or URN (ie. `urn:uuid:3f9eaf9e-cdb0-45cc-8ecb-0e5b2bfb0c20`)
+  # format.
+  def decode(value : String)
+    case value.size
```

Comment:
I suppose we could speed up the decode part, by using loops and counters instead of arrays, but this can be done later.